src/CodeGen_C.cpp

```diff
@@ -526,14 +555,22 @@ class CppVector {
         }
     }
 
-    static Vec shuffle(const Vec &a, const int32_t indices[Lanes]) {
+    template<int N, int First, int Second, int... Rest>
+    static void shuffle_impl(Vec &dst, const Vec &src) {
+        dst.elements[N] = src.elements[First < 0 ? 0 : First];
+        shuffle_impl<N + 1, Second, Rest...>(dst, src);
```

Comment:
You don't need `Second`, you can just do Rest...

Reply:
I don't understand. Surely that would drop `Second` on the floor?

Reply:
You don't need the Second template parameter either:

    template<int N, int First, int... Rest>
    static void shuffle_impl(Vec &dst, const Vec &src) {
        dst.elements[N] = src.elements[First < 0 ? 0 : First];
        shuffle_impl<N + 1, Rest...>(dst, src);
    }

Should work.

Reply:
No, if I omit that, the code become ambiguous, since `shuffle_impl<1>()` can match either of the expansions, since `...Rest` can be empty.

Reply:
Ah, the other thing I did differently in similar code was to have the base case be empty, not the last element. That avoids duplicating the body of the recursion.

Reply:
ahh, ok, that makes sense, let me try that.